### PR TITLE
Support multiple versions of header

### LINKED ISSUE #46: Add an integration test based on the `espresso-dev-node`

### PR DESCRIPTION
**Pull request by ImJeremyHe**
Closes #46 #50 

Currently, this Go client only can parse the legacy headers(v0.1) and espresso sequencer has 3 kinds of headers already. And supporting these headers is necessary for users who want to try the marketplace funcationality with the `espresso-dev-node`

### This PR:
- Define the Header of the verison 0.3
- Support the serde of multiple versions
- Test the client API with espresso dev ndoe
- Update tests
- Restruct the folders @Sneh1999 

### Key places to review:
- `HeaderImpl` in `headers.go`

### CODE REVIEW

**Review by Sneh1999 — COMMENTED** [review]
@ImJeremyHe @sveitser @zacshowa  Let me know your thoughts on the same but I was thinking maybe we should structure our repo similar to what we have in the types crate.

It has a [v0](https://github.com/EspressoSystems/espresso-sequencer/tree/main/types/src) directory  and within v0 there are multiple minor versions

Header is represented as an [enum](https://github.com/EspressoSystems/espresso-sequencer/blob/main/types/src/v0/header.rs#L12)

**Review by Sneh1999 — COMMENTED** [review]
Created a folder structure like:
```
types
  |--- common
  |--- v0
           |----v0_1
           |-----v0_3
  |----header
  ```

File: types/v0/v0_1/chain_config.go [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+}
+
+type EitherChainConfig struct {
+	Left  *ChainConfig   `json:"Left"`
+	Right **TaggedBase64 `json:"Right"`
+}
+
+func (i *EitherChainConfig) UnmarshalJSON(b []byte) error {
+	type Dec struct {
+		Left  *ChainConfig   `json:"Left"`
```
Comment (Sneh1999) [inline types/v0/v0_1/chain_config.go:62]: nit: can we lower case this ` `json:"left"` and `json:"right"` ?

File: types/v0/v0_1/chain_config.go [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+
+	if i.Left == nil && i.Right == nil {
+		return fmt.Errorf("either Left or Right variant for EitherChainConfig is required")
+	}
+
+	return nil
+}
+
+func (i *EitherChainConfig) MarshalJSON() ([]byte, error) {
+	type Left struct {
```
Comment (Sneh1999) [inline types/v0/v0_1/chain_config.go:85]: same here

File: types/v0/v0_3/chain_config.go [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+}
+
+type EitherChainConfig struct {
+	Left  *ChainConfig   `json:"Left"`
+	Right **TaggedBase64 `json:"Right"`
+}
+
+func (i *EitherChainConfig) UnmarshalJSON(b []byte) error {
+	type Dec struct {
+		Left  *ChainConfig   `json:"Left"`
```
Comment (Sneh1999) [inline types/v0/v0_3/chain_config.go:69]: same here

File: types/v0/v0_3/chain_config.go [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+	if i.Left == nil && i.Right == nil {
+		return fmt.Errorf("either Left or Right variant for EitherChainConfig is required")
+	}
+
+	return nil
+}
+
+func (i *EitherChainConfig) MarshalJSON() ([]byte, error) {
+	type Left struct {
+		Left *ChainConfig `json:"Left"`
```
Comment (Sneh1999) [inline types/v0/v0_3/chain_config.go:93]: same here

**Review by Sneh1999 — COMMENTED** [review]
Good Work @ImJeremyHe

File: types/v0/v0_3/chain_config.go (hunk 4 above)
Comment (Sneh1999) [reply types/v0/v0_3/chain_config.go]: https://github.com/EspressoSystems/espresso-sequencer-go/pull/52/commits/99bd1f5434d38edc2d9524cefb575e52ffba0453

File: types/v0/v0_3/chain_config.go (hunk 3 above)
Comment (Sneh1999) [reply types/v0/v0_3/chain_config.go]: https://github.com/EspressoSystems/espresso-sequencer-go/pull/52/commits/99bd1f5434d38edc2d9524cefb575e52ffba0453

File: types/v0/v0_1/chain_config.go (hunk 2 above)
Comment (Sneh1999) [reply types/v0/v0_1/chain_config.go:85]: https://github.com/EspressoSystems/espresso-sequencer-go/pull/52/commits/99bd1f5434d38edc2d9524cefb575e52ffba0453

File: types/v0/v0_1/chain_config.go (hunk 1 above)
Comment (Sneh1999) [reply types/v0/v0_1/chain_config.go]: https://github.com/EspressoSystems/espresso-sequencer-go/pull/52/commits/99bd1f5434d38edc2d9524cefb575e52ffba0453

File: .prettierrc.yml [hunk 5]
```diff
@@ -0,0 +1,7 @@
+bracketSpacing: true
```
[the remark this answers is not in the document]
Comment (Sneh1999) [reply .prettierrc.yml]: https://github.com/EspressoSystems/espresso-sequencer-go/pull/52/commits/99bd1f5434d38edc2d9524cefb575e52ffba0453

**Review by zacshowa — APPROVED** [review]
Tests are passing and the code looks good to me.

File: types/header.go [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+type Version = common_types.Version
+type VidCommon = common_types.VidCommon
+
+type TransactionQueryData = common_types.TransactionQueryData
+type VidCommonQueryData = common_types.VidCommonQueryData
+
+type U256 = common_types.U256
+type U256Decimal = common_types.U256Decimal
+
+var NewU256 = common_types.NewU256
```
[the remark this answers is not in the document]
Comment (ImJeremyHe) [reply types/header.go:50]: Just want to keep the interface stable for the users

**Pull request merged**